### PR TITLE
feat: add SGX driver installation on C-series VMs

### LINKED ISSUE #317: Enable SGX driver installation on C-series VMs
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
FEATURE REQUEST
---

**What version of aks-engine?**:
N/A
---

Microsoft Azure offers Intel SGX-enabled VMs (C-series) for confidential computing.
In order to run these VMs in an AKS cluster, SGX driver has to be installed on the agent nodes.

### PR DESCRIPTION
**Pull request by dmitsh**
**Reason for Change**:
Added new ACC (Azure Confidential Compute) distro based on Ubuntu 16.04 image.
Added Intel SGX driver installation on C-series VMs.

**Issue Fixed**:

Fixes #317

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jackfrancis** [comment]
thx @dmitsh!

File: parts/k8s/kubernetesinstalls.sh [hunk 1]
```diff
@@ -53,6 +53,35 @@ installGPUDrivers() {
...
+    wait_for_apt_locks
+    retrycmd_if_failure 30 5 3600 apt-get -y install $PACKAGES
+
+    local SGX_DRIVER=$(basename $SGX_DRIVER_URL)
+    local OE_DIR=/opt/azure/containers/oe
+    mkdir -p ${OE_DIR}
+
+    retrycmd_if_failure 120 5 25 curl -fsSL ${SGX_DRIVER_URL} -o ${OE_DIR}/${SGX_DRIVER}
+    chmod a+x ${OE_DIR}/${SGX_DRIVER}
+    ${OE_DIR}/${SGX_DRIVER}
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply parts/k8s/kubernetesinstalls.sh]: @dmitsh sorry, I added this comment last week but github UI fooled me and it never submitted. This is my only comment. Basically, let's add a distinct error code (see `parts/k8s/kubernetesprovisionsource.sh`) to identify when the execution of the driver fails.

File: parts/k8s/kubernetesinstalls.sh [hunk 2]
```diff
@@ -53,6 +53,35 @@ installGPUDrivers() {
...
+        ;;
+    "*")
+        echo "Version $VERSION is not supported"
+        exit 1
+        ;;
+    esac
+
+    PACKAGES="make gcc dkms"
+    wait_for_apt_locks
+    retrycmd_if_failure 30 5 3600 apt-get -y install $PACKAGES
```
Comment (jackfrancis) [inline parts/k8s/kubernetesinstalls.sh]: Also, let's add an error code like `ERR_SGX_DRIVERS_INSTALL_TIMEOUT` and replace the line above with:

`retrycmd_if_failure 30 5 3600 apt-get -y install $PACKAGES || exit $ERR_SGX_DRIVERS_INSTALL_TIMEOUT`

File: parts/k8s/kubernetesinstalls.sh [hunk 3]
```diff
@@ -53,6 +53,35 @@ installGPUDrivers() {
...
+
+    local SGX_DRIVER=$(basename $SGX_DRIVER_URL)
+    local OE_DIR=/opt/azure/containers/oe
+    mkdir -p ${OE_DIR}
+
+    retrycmd_if_failure 120 5 25 curl -fsSL ${SGX_DRIVER_URL} -o ${OE_DIR}/${SGX_DRIVER}
```
Comment (jackfrancis) [inline parts/k8s/kubernetesinstalls.sh]: Replace with `retrycmd_if_failure 120 5 25 curl -fsSL ${SGX_DRIVER_URL} -o ${OE_DIR}/${SGX_DRIVER} || exit $ERR_SGX_DRIVERS_INSTALL_TIMEOUT`

File: parts/k8s/kubernetesinstalls.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackfrancis) [reply parts/k8s/kubernetesinstalls.sh]: This error code could be identified by `ERR_SGX_DRIVERS_START_FAIL`.

So, in summary, let's add two vars to `parts/k8s/kubernetesprovisionsource.sh`:

```
ERR_SGX_DRIVERS_INSTALL_TIMEOUT=90 # Timeout waiting for SGX prereqs to download
ERR_SGX_DRIVERS_START_FAIL=91 # Failed to execute SGX driver binary
```

(90 and 91 are available as of now)

Thanks!

**Comment by tariq1890** [comment]
Once review comments are addressed. I'll implement these changes in my branch and we should be able to merge this PR then :)

**Comment by dmitsh** [comment]
Thanks @jackfrancis ! Addressed.

**Comment by dmitsh** [comment]
Thanks @jackfrancis , @tariq1890 .
With the latest changes I've successfully deployed and validated SGX cluster.
The error in AKS Engine VHD Pipeline test seems to be irrelevant - a storage account access error?

**Comment by jackfrancis** [comment]
lgtm

@tariq1890 feel free to incorporate into your fork and then merge

**Comment by CecileRobertMichon** [comment]
Sorry I'm late to the party here but I was looking this change that I missed last week and trying to solve a merge conflict with my PR and I have a few questions:
- Is this new distro/feature meant to be used by AKS-Engine users? If so can you add it to the documentation in a follow-up PR @dmitsh? 
- Any C-series VM SKUs _need_ to use the `ACC1604` distro correct? And vice-versa, that distro needs to be used with C series? If so, we should add validation to ensure that an apimodel with one or the other missing fails validation.
- Just a small observation: the `if $FULL_INSTALL_REQUIRED` for installing the driver in custom script is irrelevant because that is only used with the AKS golden image, if using the ACC1604 distro image no software will be pre-installed so `FULL_INSTALL_REQUIRED` is always true.

**Comment by dmitsh** [comment]
Thank you @CecileRobertMichon . I will send another PR.
To your question about VM SKU, ACC1604 is a special distro for Ubuntu 16.04 that supports UEFI, which is used on Gen-2 VMs, which in turn supports SGX driver. So yes, we have to use ACC1604 for C-series and wise versa.
For Ubuntu 18.04 we will be using the standard image from the Azure Image Gallery, as it already supports UEFI.

**Comment by jjcollinge** [comment]
@dmitsh I see this PR enables the provisioning of ACC machines but doesn't appear to include any modifications to the vanilla Kubernetes distro to make it EPC aware. What is the expected behavior of the Kubernetes scheduler for SGX workloads? What happens when the EPC is full but Kubernetes doesn't know it?

**Comment by jjcollinge** [comment]
To answer my own question ^. Looks like this will be made available by a plugin running as a daemonset: https://azure.microsoft.com/en-us/blog/bringing-confidential-computing-to-kubernetes/

**Pull request merged**